### PR TITLE
Fix issues with validation of hidden fields

### PR DESCRIPTION
**Pull request by Peterburnett**
Previously all fields would be marked as required, even ones not possible to set as they're hidden. This is a deeper bug exposed by previous changes in #8 , which replaced all instances of '@' with an empty string, as they were exposed in places that didnt load via the customfields API, such as the profile fields splash page. Now the fields are truly empty, but still required in the form when they ARE to be set.

### CODE REVIEW

File: field.class.php [hunk 1]
```diff
@@ -14,6 +14,8 @@
 // You should have received a copy of the GNU General Public License
 // along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
+use engage_survey\result\option;
```
Comment (keevan) [inline field.class.php]: Weird use statement :)

File: field.class.php [hunk 2]
```diff
@@ -461,9 +461,33 @@ public function edit_field_set_required($mform) {
...
+                    $infodata = $DB->get_record('user_info_data', $params);
+                    if (!$infodata->data) {
+                        // Always required in a completely blank form.
+                        $req = true;
+                    } else if ($infodata->data === $this->field->param4) {
+                        $req = true;
+                    }
+                } else if($parentval === $this->field->param4) {
+                    $req = true;
+                }
```
Comment (keevan) [inline field.class.php:487]: Though this has been merged I wanted to do a proper review on this point:

- This should probably be moved to its own function, probably a private helper method, perhaps something like `$this->some_field_is_required()`
- `$req` should be `$required` to disambiguate it from "request or required"
- Nested if statements are hard to follow

I zoomed out a bit and saw how difficult to follow this whole function appears to be.

Comment (Peterburnett) [reply field.class.php:487]: I agree the logic here is tough. Its a bit easier if you have a complete grip on the plugin, but its not great. I didnt split it to its own function purely because it is already nested inside of a specific function for settting 'required' so the extra abtraction didnt feel entirely necessary.
